### PR TITLE
MediaPlayer WebAssembly

### LINKED ISSUE #11967: MediaPlayer WebAssembly
This issue is an epic to the support for MediaPlayerElement on WebAssembly.

## Blocking issues
- [x] #12531
- [ ] #12533
- [ ] #12535
- [x] WebAssembly fullscreen is not working property
- [x] The transport control flickers when clicking any buttons - PR: https://github.com/unoplatform/uno/pull/12372
- [x] https://github.com/unoplatform/uno/issues/12437
- Migrate [this table](https://docs.google.com/spreadsheets/d/135nS-P1XeaLVnTvtDl69lDzAfXcGHuN8-RpLdabhmD0/edit#gid=0) to a markdown table in this issue.
- We need to take the table to the docs 
- - https://platform.uno/docs/articles/controls/MediaPlayerElement.html
- - https://github.com/unoplatform/uno/blob/master/doc/articles/controls/MediaPlayerElement.md
- But we need to finish the tests for each features and Media formats
- [x] https://github.com/unoplatform/uno/issues/12350 - PR https://github.com/unoplatform/uno/pull/12372

 ## Gallery samples issues
 - [x] (Rafael) https://github.com/unoplatform/uno/issues/12330
 	^ PR: https://github.com/unoplatform/uno/pull/12398
 - [ ] (Agnès) https://github.com/unoplatform/uno/issues/12331
 - [ ] (Agnès) https://github.com/unoplatform/uno/issues/12332
 - [x] (Xiaoy) https://github.com/unoplatform/uno/issues/12344
 	^ PR: #12440
 - [x] (Rafael) https://github.com/unoplatform/uno/issues/12352
 	^ PR: https://github.com/unoplatform/uno/pull/12416
 - [x] (Rafael) https://github.com/unoplatform/uno/issues/12338
 - [x] (Rafael) https://github.com/unoplatform/uno/issues/12341
 - [x] (Xiaoy) https://github.com/unoplatform/uno/issues/12333
 	^ PR: #12440
 - [x] (Xiaoy) https://github.com/unoplatform/uno/issues/12335
 	^ PR: #12440
 - [x] (Rafael) https://github.com/unoplatform/uno/issues/12499 - ^ PR: #12500 
 
## Future Features
- [ ] (Rafael) https://github.com/unoplatform/uno/issues/12336 - PR: https://github.com/unoplatform/uno/pull/12404
- [ ] https://github.com/unoplatform/uno/issues/12337
- [ ] https://github.com/unoplatform/uno/issues/12394
- [ ] ZoomButton need to show all option - Aspect - https://learn.microsoft.com/en-us/xamarin/community-toolkit/views/mediaelement
- [ ] Repeat show if true or false: show different icon or border state?
- [ ] FastRewind or FastForward need to show the rate? (1x, 1.25x or 3x, etc)
- [ ] Download option
- [ ] Speed
- [x] Picture-in-picture - Like https://www.w3schools.com/html/html5_video.asp
- [ ] Update the Visual of the Wasm to look like the Android (Pixel Perfect)
- [ ] Make the Media run playlist (Files m3u8) (Using [VideoJS](https://videojs.com/))
- [ ] Make the style customizable
- [ ] Should we set a max (or min for rewind)?
 - [ ] Update the Visual of the Wasm to look like the Android (Pixel Perfect)
 - [ ] Make the Media run playlist (Files m3u8) (Using [VideoJS](https://videojs.com/))

### PR DESCRIPTION
**Pull request by rafael-rosa-knowcode**
GitHub Issue (If applicable): 
Issue: #11967 

## PR Type

What kind of change does this PR introduce?

- Feature

## What is the current behavior?

The Media Player Extensions do not work on Wasm

## What is the new behavior?

Now we have a Media Player Extensions for Wasm

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/AddIns/Uno.UI.MediaPlayer.WebAssembly/HtmlMediaPlayer.NativeMethods.cs [hunk 1]
```diff
@@ -5,7 +5,9 @@
 #endif
 
 using System;
+#if NET7_0_OR_GREATER
```
Comment (nickrandolph) [inline src/AddIns/Uno.UI.MediaPlayer.WebAssembly/HtmlMediaPlayer.NativeMethods.cs:8]: Is this not supported on NET6?

File: src/AddIns/Uno.UI.MediaPlayer.WebAssembly/HtmlMediaPlayer.cs [hunk 2]
```diff
@@ -8,6 +8,8 @@
 using Windows.Media.Playback;
 using Windows.UI.Xaml.Controls;
 using Windows.UI.Xaml;
+using Windows.UI.Xaml.Controls.Maps;
```
Comment (nickrandolph) [inline src/AddIns/Uno.UI.MediaPlayer.WebAssembly/HtmlMediaPlayer.cs:11]: Why are we importing Maps namespace?

File: src/AddIns/Uno.UI.MediaPlayer.WebAssembly/HtmlMediaPlayer.cs [hunk 3]
```diff
@@ -476,4 +478,83 @@ public void Stop()
...
 	}
+
+	public double _playbackRate;
+	public double PlaybackRate
+	{
+		get => _playbackRate;
+		set
+		{
+			_playbackRate = value;
+			if (IsVideo)
```
Comment (nickrandolph) [inline src/AddIns/Uno.UI.MediaPlayer.WebAssembly/HtmlMediaPlayer.cs:489]: Instead of having all these "IsVideo"/"IsAudio" logic, can we have a simple property that wraps this eg
ActiveElement => IsVideo? _htmlVideo:
                                    IsAudio? _htmlVideo: 
                                             default;

File: src/AddIns/Uno.UI.MediaPlayer.WebAssembly/HtmlMediaPlayer.cs [hunk 4]
```diff
@@ -476,4 +478,83 @@ public void Stop()
...
+		set
+		{
+			_isLoopingEnabled = value;
+			if (_isLoopingEnabled)
+			{
+				if (IsVideo)
+				{
+					if (!string.IsNullOrEmpty(_htmlVideo.GetHtmlAttribute("loop")))
+					{
+						_htmlVideo.SetHtmlAttribute("loop", "");
```
Comment (nickrandolph) [inline src/AddIns/Uno.UI.MediaPlayer.WebAssembly/HtmlMediaPlayer.cs:512]: Why are we setting loop to "" here?

File: src/AddIns/Uno.UI.MediaPlayer.WebAssembly/MediaPlayerExtension.cs [hunk 5]
```diff
@@ -328,6 +364,7 @@ public void Play()
 
 			if (_isPlayerPrepared)
 			{
+				_player.PlaybackRate = 1;
```
Comment (nickrandolph) [inline src/AddIns/Uno.UI.MediaPlayer.WebAssembly/MediaPlayerExtension.cs:367]: Check PlaybackRate before defaulting to 1

File: src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/MediaPlayerElement/MediaPlayerElement_Original.xaml [hunk 6]
```diff
@@ -4,7 +4,18 @@
...
+	<!--<MediaPlayerElement x:Name="mediaPlayerElement" Source="https://test-videos.co.uk/vids/bigbuckbunny/mp4/h264/720/Big_Buck_Bunny_720_10s_5MB.mp4"-->
+	<!--<MediaPlayerElement Source="https://www.youtube.com/embed/DDtzqzIqgdw"-->
+
+	<!--<MediaPlayerElement Source="ms-appx:///Assets/Big_Buck_Bunny_1080_10s_5MB.mp4"
+		AreTransportControlsEnabled="True"
+		AutoPlay="True" />-->
+
+
+
+	<MediaPlayerElement Source="http://localhost:5000/Big_Buck_Bunny_1080_10s_5MB.mp4"
```
Comment (nickrandolph) [inline src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/MediaPlayerElement/MediaPlayerElement_Original.xaml:17]: Don't commit this url

File: src/Uno.UI/UI/Xaml/Controls/MediaPlayerElement/MediaTransportControls.MediaPlayer.cs [hunk 7]
```diff
@@ -44,10 +46,65 @@ private void BindMediaPlayer()
...
+				_mediaPlayer.Pause();
+				_mediaPlayer.PlaybackSession.Position = TimeSpan.FromSeconds(_progressSlider.Value);
+				if (_wasPlaying)
+				{
+					_mediaPlayer.Play();
+				}
+			}
+		}
+		public void DragOverProgressSlider(object sender, RoutedEventArgs e)
+		{
```
Comment (nickrandolph) [inline src/Uno.UI/UI/Xaml/Controls/MediaPlayerElement/MediaTransportControls.MediaPlayer.cs:92]: Should we be adjusting play position whilst they're dragging? I would have assumed once they release the slider we set play position

File: src/Uno.UI/UI/Xaml/Controls/MediaPlayerElement/MediaTransportControls.MediaPlayer.cs [hunk 8]
```diff
@@ -144,9 +203,46 @@ private void OnPositionChanged(MediaPlaybackSession sender, object args)
...
+				&& _mediaPlayer.PlaybackSession.UpdateTimePositionRate != 1)
+			{
+				_mediaPlayer.PlaybackSession.IsUpdateTimePosition = true;
+				elapsed += TimeSpan.FromSeconds(_mediaPlayer.PlaybackSession.UpdateTimePositionRate);
+				_mediaPlayer.PlaybackSession.Position = elapsed;
+				await Task.Delay(250);
+				_mediaPlayer.PlaybackSession.IsUpdateTimePosition = false;
+			}
+		}
+		//public void UpdatePlaybackRate(TimeSpan elapsed)
```
Comment (nickrandolph) [inline src/Uno.UI/UI/Xaml/Controls/MediaPlayerElement/MediaTransportControls.MediaPlayer.cs:224]: Remove commented code

File: src/Uno.UI/UI/Xaml/Controls/MediaPlayerElement/MediaTransportControls.MediaPlayer.cs [hunk 9]
```diff
@@ -144,9 +203,46 @@ private void OnPositionChanged(MediaPlaybackSession sender, object args)
...
+		//		&& _mediaPlayer.PlaybackSession.PlaybackRate != 0
+		//		&& _mediaPlayer.PlaybackSession.PlaybackRate != 1)
+		//	{
+		//		_mediaPlayer.PlaybackRate = _mediaPlayer.PlaybackSession.PlaybackRate;
+		//	}
+		//}
+		private void ResetProgressSlider()
+		{
+			var elapsed = TimeSpan.Zero;
+			_timeElapsedElement.Maybe(p => p.Text = $"{elapsed.TotalHours:0}:{elapsed.Minutes:00}:{elapsed.Seconds:00}");
```
Comment (nickrandolph) [inline src/Uno.UI/UI/Xaml/Controls/MediaPlayerElement/MediaTransportControls.MediaPlayer.cs:238]: Is this formatting controlled somehow?

File: src/Uno.UI/UI/Xaml/Controls/MediaPlayerElement/MediaTransportControls.MediaPlayer.cs [hunk 10]
```diff
@@ -173,6 +271,17 @@ private void SkipForward(object sender, RoutedEventArgs e)
 		{
 			_mediaPlayer.PlaybackSession.Position = _mediaPlayer.PlaybackSession.Position + TimeSpan.FromSeconds(30);
 		}
+		private void ForwardButton(object sender, RoutedEventArgs e)
+		{
+			_mediaPlayer.PlaybackSession.UpdateTimePositionRate = _mediaPlayer.PlaybackSession.UpdateTimePositionRate < 1 ? 1 : /*To stop the Rewind*/
+														_mediaPlayer.PlaybackSession.UpdateTimePositionRate * 2; /*To start the Forward*/
```
Comment (nickrandolph) [inline src/Uno.UI/UI/Xaml/Controls/MediaPlayerElement/MediaTransportControls.MediaPlayer.cs:277]: Should we set a max (or min for rewind)?

File: src/AddIns/Uno.UI.MediaPlayer.WebAssembly/HtmlMediaPlayer.NativeMethods.cs (hunk 1 above)
Comment (Youssef1313) [inline src/AddIns/Uno.UI.MediaPlayer.WebAssembly/HtmlMediaPlayer.NativeMethods.cs:8]:
```suggestion
#if USE_JSIMPORT
```

File: src/AddIns/Uno.UI.MediaPlayer.WebAssembly/HtmlMediaPlayer.cs [hunk 11]
```diff
@@ -476,4 +478,83 @@ public void Stop()
 
 		NativeMethods.Stop(IsAudio ? _htmlAudio.HtmlId : _htmlVideo.HtmlId);
 	}
+
+	public double _playbackRate;
```
Comment (Youssef1313) [inline src/AddIns/Uno.UI.MediaPlayer.WebAssembly/HtmlMediaPlayer.cs:482]: private?

File: src/AddIns/Uno.UI.MediaPlayer.WebAssembly/HtmlMediaPlayer.cs [hunk 12]
```diff
@@ -476,4 +478,83 @@ public void Stop()
...
+	public double PlaybackRate
+	{
+		get => _playbackRate;
+		set
+		{
+			_playbackRate = value;
+			if (IsVideo)
+			{
+				_htmlVideo.ExecuteJavascript("element.playbackRate = '" + _playbackRate.ToString("0.00", CultureInfo.InvariantCulture) + "';");
```
Comment (Youssef1313) [inline src/AddIns/Uno.UI.MediaPlayer.WebAssembly/HtmlMediaPlayer.cs:491]: @jeromelaban @ebariche Is this the right way?

File: src/AddIns/Uno.UI.MediaPlayer.WebAssembly/HtmlMediaPlayer.cs [hunk 13]
```diff
@@ -476,4 +478,83 @@ public void Stop()
...
+			}
+			else if (IsAudio)
+			{
+				_htmlAudio.ExecuteJavascript("element.playbackRate = '" + _playbackRate.ToString("0.00", CultureInfo.InvariantCulture) + "';");
+			}
+		}
+	}
+	public bool _isLoopingEnabled;
```
Comment (Youssef1313) [inline src/AddIns/Uno.UI.MediaPlayer.WebAssembly/HtmlMediaPlayer.cs:499]: private?

File: src/AddIns/Uno.UI.MediaPlayer.WebAssembly/MediaPlayerExtension.cs [hunk 14]
```diff
@@ -102,12 +106,44 @@ void IMediaPlayerExtension.OnOptionChanged(string name, object value)
 
 	private void ApplyAnonymousCors()
 		=> _player?.SetAnonymousCORS(_anonymousCors);
+	private void ApplyPlaybackRate()
```
Comment (Youssef1313) [inline src/AddIns/Uno.UI.MediaPlayer.WebAssembly/MediaPlayerExtension.cs:109]:
```suggestion

	private void ApplyPlaybackRate()
```

File: src/AddIns/Uno.UI.MediaPlayer.WebAssembly/MediaPlayerExtension.cs [hunk 15]
```diff
@@ -102,12 +106,44 @@ void IMediaPlayerExtension.OnOptionChanged(string name, object value)
...
+		}
+	}
 
 	public IMediaPlayerEventsExtension? Events { get; set; }
 
-	public double PlaybackRate { get; set; }
+	//public double PlaybackRate { get; set; }
 
-	public bool IsLoopingEnabled { get; set; }
+	public double _playbackRate;
```
Comment (Youssef1313) [inline src/AddIns/Uno.UI.MediaPlayer.WebAssembly/MediaPlayerExtension.cs:121]: private?

File: src/AddIns/Uno.UI.MediaPlayer.WebAssembly/MediaPlayerExtension.cs [hunk 16]
```diff
@@ -102,12 +106,44 @@ void IMediaPlayerExtension.OnOptionChanged(string name, object value)
...
+		set
+		{
+			_playbackRate = value;
+			if (_player is not null)
+			{
+				_player.PlaybackRate = value;
+			}
+		}
+	}
+	public bool _isLoopingEnabled;
```
Comment (Youssef1313) [inline src/AddIns/Uno.UI.MediaPlayer.WebAssembly/MediaPlayerExtension.cs:134]: private?

File: src/Uno.UI/UI/Xaml/Controls/MediaPlayerElement/MediaTransportControls.MediaPlayer.cs [hunk 17]
```diff
@@ -44,10 +46,65 @@ private void BindMediaPlayer()
...
+			_fastForwardButton.Maybe(p => p.Tapped += ForwardButton);
+			_rewindButton.Maybe(p => p.Tapped += RewindButton);
+			_progressSlider.Maybe(p => p.Tapped += TappedProgressSlider);
+			_progressSlider.Maybe(p => p.DragLeave += DragLeaveProgressSlider);
+			_progressSlider.Maybe(p => p.DragOver += DragOverProgressSlider);
 
 			_subscriptions.Disposable = AttachThumbEventHandlers(_progressSlider);
 		}
-
+		public void TappedProgressSlider(object sender, RoutedEventArgs e)
```
Comment (Youssef1313) [inline src/Uno.UI/UI/Xaml/Controls/MediaPlayerElement/MediaTransportControls.MediaPlayer.cs:57]:
```suggestion

		public void TappedProgressSlider(object sender, RoutedEventArgs e)
```

File: src/Uno.UI/UI/Xaml/Controls/MediaPlayerElement/MediaTransportControls.cs [hunk 18]
```diff
@@ -102,18 +104,14 @@ public partial class MediaTransportControls : Control
 		private bool _wasPlaying;
 		private bool _isInteractive;
 		private MediaPlayerElement _mpe;
-
 		public MediaTransportControls() : base()
```
Comment (Youssef1313) [inline src/Uno.UI/UI/Xaml/Controls/MediaPlayerElement/MediaTransportControls.cs:107]:
```suggestion

		public MediaTransportControls() : base()
```

File: src/Uno.UWP/Media/Playback/MediaPlaybackSession.cs [hunk 19]
```diff
@@ -63,6 +67,26 @@ internal set
...
+
+		private double _updateTimePositionRate;
+		public double UpdateTimePositionRate
+		{
+			get => _updateTimePositionRate;
+			set
+			{
+				_updateTimePositionRate = value;
+			}
+		}
```
Comment (Youssef1313) [inline src/Uno.UWP/Media/Playback/MediaPlaybackSession.cs:88]:
```suggestion
		public bool IsUpdateTimePosition { get; set; }

		public double UpdateTimePositionRate { get; set; }
```

File: src/Uno.UI/UI/Xaml/Controls/MediaPlayerElement/MediaTransportControls.cs [hunk 20]
```diff
@@ -259,6 +273,24 @@ private void FullWindowButtonTapped(object sender, RoutedEventArgs e)
...
+		private void PreviousTrackButtonTapped(object sender, RoutedEventArgs e)
+		{
+			_mediaPlayer.PlaybackSession.Position = TimeSpan.Zero;
+			_mpe.MediaPlayer.PlaybackSession.Position = TimeSpan.Zero;
+		}
+		private void NextTrackButtonTapped(object sender, RoutedEventArgs e)
+		{
+			_mediaPlayer.PlaybackSession.Position = _mediaPlayer.PlaybackSession.NaturalDuration;
+			_mpe.MediaPlayer.PlaybackSession.Position = _mediaPlayer.PlaybackSession.NaturalDuration;
+		}
```
Comment (Youssef1313) [inline src/Uno.UI/UI/Xaml/Controls/MediaPlayerElement/MediaTransportControls.cs:293]: Leave empty line between methods

File: src/AddIns/Uno.UI.MediaPlayer.WebAssembly/HtmlMediaPlayer.cs (hunk 12 above)
Comment (ebariche) [reply src/AddIns/Uno.UI.MediaPlayer.WebAssembly/HtmlMediaPlayer.cs:491]: No

File: src/AddIns/Uno.UI.MediaPlayer.WebAssembly/HtmlMediaPlayer.cs [hunk 21]
```diff
@@ -476,4 +478,83 @@ public void Stop()
...
+		set
+		{
+			_playbackRate = value;
+			if (IsVideo)
+			{
+				_htmlVideo.ExecuteJavascript("element.playbackRate = '" + _playbackRate.ToString("0.00", CultureInfo.InvariantCulture) + "';");
+			}
+			else if (IsAudio)
+			{
+				_htmlAudio.ExecuteJavascript("element.playbackRate = '" + _playbackRate.ToString("0.00", CultureInfo.InvariantCulture) + "';");
```
Comment (ebariche) [inline src/AddIns/Uno.UI.MediaPlayer.WebAssembly/HtmlMediaPlayer.cs:495]: Neither is this.

File: src/AddIns/Uno.UI.MediaPlayer.WebAssembly/HtmlMediaPlayer.NativeMethods.cs [hunk 22]
```diff
@@ -55,6 +57,16 @@ internal static partial class NativeMethods
...
 #endif
 
+#if USE_JSIMPORT
+		[JSImport("globalThis.Uno.UI.Media.HtmlMediaPlayer.setPlaybackRate")]
+#endif
+		internal static partial void SetPlaybackRate(nint htmlId, double playbackRate);
+
+#if !USE_JSIMPORT
+		internal static partial void SetPlaybackRate(nint htmlId, double playbackRate)
+			=> throw new NotSupportedException();
```
Comment (ebariche) [inline src/AddIns/Uno.UI.MediaPlayer.WebAssembly/HtmlMediaPlayer.NativeMethods.cs:67]: @jeromelaban : Are we supporting < NET7 ?

**Review by ebariche — COMMENTED** [review]
image

File: src/AddIns/Uno.UI.MediaPlayer.WebAssembly/HtmlMediaPlayer.cs [hunk 23]
```diff
@@ -474,6 +431,39 @@ public void Stop()
...
+			{
+				this.Log().Debug($"{ActiveElementName} loop: [{Source}]");
+			}
+		}
+		else
+		{
+			ActiveElement.ClearHtmlAttribute("loop");
+			if (this.Log().IsEnabled(Uno.Foundation.Logging.LogLevel.Debug))
+			{
+				this.Log().Debug($"{ActiveElementName} loop: [{Source}]");
```
Comment (ebariche) [inline src/AddIns/Uno.UI.MediaPlayer.WebAssembly/HtmlMediaPlayer.cs:465]: Log once or change message

File: src/SamplesApp/SamplesApp.Wasm/SamplesApp.Wasm.csproj [hunk 24]
```diff
@@ -16,6 +16,7 @@
 		<UseUnoXamlParser>true</UseUnoXamlParser>
 		<UnoRuntimeIdentifier>WebAssembly</UnoRuntimeIdentifier>
 		<WasmShellIncludeWindowsCompatibility>false</WasmShellIncludeWindowsCompatibility>
+		<WasmShellEnableThreads>false</WasmShellEnableThreads>
```
Comment (jeromelaban) [inline src/SamplesApp/SamplesApp.Wasm/SamplesApp.Wasm.csproj:19]: This is not needed (Disabled by default)

File: src/SamplesApp/SamplesApp.Wasm/Program.cs [hunk 25]
```diff
@@ -19,7 +19,7 @@ public static void Main(string[] args)
 #endif
 
 			// Always enable anonymous videos for Samples App.
-			Uno.UI.Media.FeatureConfiguration.AnonymousCorsDefault = true;
+			Uno.UI.Media.FeatureConfiguration.AnonymousCorsDefault = false;
```
Comment (jeromelaban) [inline src/SamplesApp/SamplesApp.Wasm/Program.cs:22]: This can be removed (false is the default)

File: src/AddIns/Uno.UI.MediaPlayer.WebAssembly/MediaPlayerExtension.cs [hunk 26]
```diff
@@ -228,7 +228,6 @@ private void InitializePlayer()
 		_owner.PlaybackSession.PlaybackStateChanged -= OnStatusChanged;
 		_owner.PlaybackSession.PlaybackStateChanged += OnStatusChanged;
 
-		ApplyAnonymousCors();
```
Comment (jeromelaban) [inline src/AddIns/Uno.UI.MediaPlayer.WebAssembly/MediaPlayerExtension.cs:210]: Why is this removed? If someone sets the default, it may not be applied properly.

File: src/AddIns/Uno.UI.MediaPlayer.WebAssembly/HtmlMediaPlayer.NativeMethods.cs (hunk 22 above)
Comment (jeromelaban) [reply src/AddIns/Uno.UI.MediaPlayer.WebAssembly/HtmlMediaPlayer.NativeMethods.cs:67]: We're not.

File: src/AddIns/Uno.UI.MediaPlayer.WebAssembly/HtmlVideo.cs [hunk 27]
```diff
@@ -0,0 +1,15 @@
...
+
+namespace Uno.UI.Media;
+
+internal class HtmlVideo : UIElement
+{
+	public HtmlVideo() : base("video")
+	{
+		SetAttribute("background-color", "transparent");
+		SetStyle("width", "100%");
+		SetStyle("height", "100%");
```
Comment (carldebilly) [inline src/AddIns/Uno.UI.MediaPlayer.WebAssembly/HtmlVideo.cs:13]:
```suggestion
		SetStyle(
			("width", "100%"),
			("height", "100%"));
```

File: src/Uno.UWP/Generated/3.0.0.0/Windows.Media.Core/MediaSource.cs [hunk 28]
```diff
@@ -2,8 +2,8 @@
 #pragma warning disable 114 // new keyword hiding
 namespace Windows.Media.Core
 {
-	#if false || false || NET461 || __WASM__ || __SKIA__ || __NETSTD_REFERENCE__ || false
-	[global::Uno.NotImplemented("NET461", "__WASM__", "__SKIA__", "__NETSTD_REFERENCE__")]
+	#if false || false || false || false
```
[the remark this answers is not in the document]
Comment (jeromelaban) [reply src/Uno.UWP/Generated/3.0.0.0/Windows.Media.Core/MediaSource.cs:5]: It is, all targets will be supporting it.

File: src/Uno.UI/UI/Xaml/Style/Generic/Generic.xaml [hunk 29]
```diff
@@ -7075,12 +7075,12 @@
 											<Button x:Name='CCSelectionButton'
 													Style='{StaticResource MTC_ButtonStyle}'
 													Visibility='Collapsed'>
-												<FontIcon Glyph="&#xED1E;" />
+												<FontIcon FontFamily="Segoe MDL2 Assets" Glyph="&#xED1E;" />
```
[the remark this answers is not in the document]
Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/Style/Generic/Generic.xaml:7078]: For the Generic.xaml file, it is.

File: src/Uno.UI/UI/Xaml/Controls/MediaPlayerElement/MediaPlayerElement.cs [hunk 30]
```diff
@@ -12,7 +12,12 @@ namespace Windows.UI.Xaml.Controls
 	[TemplatePart(Name = TransportControlsPresenterName, Type = typeof(ContentPresenter))]
 	[TemplatePart(Name = MediaPlayerPresenterName, Type = typeof(MediaPlayerPresenter))]
 	[TemplatePart(Name = LayoutRootName, Type = typeof(Grid))]
-	public partial class MediaPlayerElement : IDisposable
+	public partial class MediaPlayerElement
+#if __IOS__ || __ANDROID__ || __MACOS__
+		// To avoid causing FrameworkElement.Dispose to become virtual (and cause a breaking change),
+		// we keep the disposable for existing platforms, but we don't implement it for the other targets.
+		: IDisposable
+#endif
```
Comment (Youssef1313) [inline src/Uno.UI/UI/Xaml/Controls/MediaPlayerElement/MediaPlayerElement.cs:20]: Should we track removing the interface completely (both on MediaPlayerElement and FrameworkElement) for Uno 5?

Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/Controls/MediaPlayerElement/MediaPlayerElement.cs:20]: Indeed, it's added to the list.

File: src/Uno.UI/UI/Xaml/WindowManagerInterop.wasm.cs [hunk 31]
```diff
@@ -579,13 +587,21 @@ internal static string GetAttribute(IntPtr htmlId, string name)
 		#region ClearAttribute
 		internal static void RemoveAttribute(IntPtr htmlId, string name)
 		{
+
+#if NET7_0_OR_GREATER
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/WindowManagerInterop.wasm.cs:591]: @rafael-rosa-knowcode Was this added on purpose? @ebariche does this ring a bell?

Comment (ebariche) [reply src/Uno.UI/UI/Xaml/WindowManagerInterop.wasm.cs:591]: No and it won't work !

image

Comment (ebariche) [reply src/Uno.UI/UI/Xaml/WindowManagerInterop.wasm.cs:591]: @rafael-rosa-knowcode : Undo all changes to this file.

1 - Neither change is needed.
2 - SetAttribute needs a specialized binding to avoid array allocations.
3 - RemoveAttribute is not implemented correctly.

File: src/Uno.UI.Runtime.WebAssembly/Xaml/UIElementWasmExtensions.cs [hunk 32]
```diff
@@ -113,6 +113,14 @@ public static void ClearHtmlAttribute(this UIElement element, string name)
 			WindowManagerInterop.RemoveAttribute(element.HtmlId, name);
 		}
 
+		/// <summary>
+		/// Clear/remove a HTML attribute from an element.
```
Comment (agneszitte) [inline src/Uno.UI.Runtime.WebAssembly/Xaml/UIElementWasmExtensions.cs:117]:
```suggestion
		/// Clear/remove an HTML attribute from an element.
```
